src/systems/systems/src/hydraulic/linear_actuator.rs

```diff
@@ -70,51 +84,87 @@ struct CoreHydraulicForce {
 
     last_control_force: Force,
 
-    force: Force,
+    force_raw: Force,
+    force_filtered: LowPassFilter<Force>,
+    max_force: Force,
+
+    pid_controller: PidController,
 }
 impl CoreHydraulicForce {
```

Comment:
This piece of code does a lot. I can see that you realize this since you describe that this aims to encapsulate - what in reality - consists of many different components. In fact, it looks like you have already created this struct simply to separate some of the internal physics of the `LinearActuator` in this, which is good.

However, I believe this can be optimized a bit. Part of the reason why this might be hard to understand is that you're supporting four different modes from `LinearActuatorMode` and dynamic transitions between them. I cannot give you a concrete suggestion as to how to optimize this right now, but if we can somehow separate those four modes, that would be a step in the right direction.

Reply:
Yeah this is a difficult part... Initially I didn't have any mode transition in there. But added the closed circuit mode, and to use it properly any user would have to know that they need to go to that mode if pressure is low.

Adding the transition here hides the required logic, and if you need to code a door controller, you don't have to know that when there's too low pressure, you'd have to switch to "closed circuit" mode.

That said, if I can find a magic formula the closed circuit mode might disapear again and everything could be handled by the closed loop control. We'll see if brain gets illuminated some day....   But user code in the controller wouldn't have to be changed though.

Modes are also really intricated as for example I need to save the last force output to start from here in the next mode to keep continuity and coherency...  Welcome in the core of the sorcery!